src/board.rs

```diff
@@ -45,13 +41,123 @@ impl Chessboard {
+        if !Self::valid_fen(fen) {
+            println!("{} is an invalid FEN. Try again.", fen);
+            chessboard
+        } else {
+            // Split the FEN string into parts using ' ' as the delimiter
+            let fen_parts: Vec<&str> = fen.split_whitespace().collect();
+
+            // Parse the piece placement part of the FEN string
+            let board_rows: Vec<&str> = fen_parts[0].split('/').collect();
+            let two: u64 = 2;
+            for (mut rank, row) in board_rows.iter().rev().enumerate() {
+                rank += 1;
+                let mut file = 0; // Initialize the file (column) index
+                for piece in row.chars() {
+                    if piece.is_digit(10) {
+                        let empty_squares = piece.to_digit(10).unwrap() as usize;
+                        file += empty_squares; // Skip empty squares
+                    } else {
+                        let square_index = 8 * (rank - 1) + file;
+                        // Update the corresponding bitboard based on the piece type and color
+                        match piece {
+                            'p' => chessboard.black_pawns |= two.pow(square_index as u32),
+                            'r' => chessboard.black_rooks |= two.pow(square_index as u32),
+                            'b' => chessboard.black_bishops |= two.pow(square_index as u32),
+                            'k' => chessboard.black_king |= two.pow(square_index as u32),
+                            'q' => chessboard.black_queen |= two.pow(square_index as u32),
+                            'n' => chessboard.black_knights |= two.pow(square_index as u32),
+                            'P' => chessboard.white_pawns |= two.pow(square_index as u32),
+                            'R' => chessboard.white_rooks |= two.pow(square_index as u32),
+                            'B' => chessboard.white_bishops |= two.pow(square_index as u32),
+                            'K' => chessboard.white_king |= two.pow(square_index as u32),
+                            'Q' => chessboard.white_queen |= two.pow(square_index as u32),
+                            'N' => chessboard.white_knights |= two.pow(square_index as u32),
+                            _ => { /* Handle other characters if needed */ }
+                        }
+                        file += 1; // Move to the next file
+                    }
+                }
+            }
+
+            // Parse whose turn it is
+            chessboard.white_turn = fen_parts[1] == "w";
+
+            // Parse castling rights
+            let fen_castle = fen_parts[2];
+            for c in fen_castle.chars() {
+                let v = match c {
+                    'K' => 0b1000,
+                    'Q' => 0b0100,
+                    'k' => 0b0010,
+                    'q' => 0b0001,
+                    _ => 0b0,
+                };
+                chessboard.castling_rights |= v;
+            }
```

Comment:
while we're at it, let's put this into a separate function too.

Reply:
Done.